### DISCUSSION TITLE
More varied image selection dialog layout

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3586] [Ideas]
<details><summary>What's the the image selection dialog?</summary>

It's the pop-up which is used by most quests that have a two-part selection. For example, in the cycleway quest, you tap a side and then there's a pop-up to pick the type of cycleway. There are some images in https://github.com/streetcomplete/StreetComplete/issues/771#issuecomment-991776256 and it appears several times in the video just above.
</details>

@westnordost, @matkoniecz, and other SC developers who have implemented quests: Right now, the image selection dialog reserves the same space for every item. When you were building the UI, have you ever wanted a more varied layout?

I'm asking because when I was thinking about the maxweight sign in https://github.com/streetcomplete/StreetComplete/issues/3557#issuecomment-989362185, I thought it would be nice to have the "not listed" option extend across the entire bottom row, something like:

<img src="https://user-images.githubusercontent.com/6563664/145726396-76f0c510-f116-45b4-9c38-c206e86ebe5e.png" alt="not listed mockup" height="350" />

I took a quick look at the source, and I think it would be too much effort to justify for just this one quest (and a "quick & dirty" solution would not be appropriate, because the `image_select` is used in many places). However, if a more varied layout is something that you've wished for in other places, then this might be a good time to implement a way to do that generically.

**Comment by matkoniecz** [comment]
> When you were building the UI, have you ever wanted a more varied layout?

No, so far I never had any use for it.

**Comment by smichel17** [comment]
For future reference— I missed this the first time around, but there is actually already is code to handle this

https://github.com/streetcomplete/StreetComplete/blob/583da6031f3b3c2ee086c7f5c9ede08c5ee283da/app/src/main/java/de/westnordost/streetcomplete/view/image_select/GroupedImageSelectAdapter.kt#L38-L42

It only handles a single special case— an item which is a group should span the whole width —but it's easy enough to add other cases there.

**Reply by smichel17** [reply]
@westnordost I noticed that in `GroupedImageSelectAdapter`, both in the lines above and `onCreateViewHolder`, we have some logic like, "If the view is a group, use one layout, otherwise use another". Would it make any sense to pull that out into the `DisplayItem`? That way each DisplayItem can define its own layout. Or is there a better way to allow something like the mockup in the original post?
